services/QuillLMS/engines/comprehension/lib/comprehension/comprehension/fetch_change_logs.rb

```diff
@@ -0,0 +1,77 @@
+module Comprehension
+  module FetchChangeLogs
+    include Comprehension::ChangeLog
+    UNIVERSAL_RULE_ACTIONS = ['Universal Rule - updated', 'Universal Rule - created']
+
+    def change_logs_for_activity(activity)
+      @activity = activity
```

Comment:
So, if we set up the Rails associations correctly, we can simplify this a lot, e.g.
```ruby
passages_change_logs = activity.passages.map(&:change_logs).flatten
```
It could also be sped up by eager loading, e.g.
```ruby 
Activity.where(id: activity_id)
  .includes(passages: [:change_logs], prompts: [:change_logs])